src/System.Text.Primitives/System/Text/Encoding/Utf16/Utf16TextEncodingLE.cs

```diff
@@ -7,113 +7,145 @@ namespace System.Text.Utf16
+
+        private const uint Utf16SurrogateRangeStart = Utf16HighSurrogateFirstCodePoint;
+        private const uint Utf16SurrogateRangeEnd = Utf16LowSurrogateLastCodePoint;
+
+        // To get this to compile with dotnet cli, we need to temporarily un-binary the magic values
+        private const byte b0000_0111U = 7;
+        private const byte b0000_1111U = 15;
+        private const byte b0001_1111U = 31;
+        private const byte b0011_1111U = 63;
+        private const byte b0111_1111U = 127;
+        private const byte b1000_0000U = 128;
+        private const byte b1100_0000U = 192;
+        private const byte b1110_0000U = 224;
+        private const byte b1111_0000U = 240;
+        private const byte b1111_1000U = 248;
+
+        #endregion Encoding Constants
+
+        #region Decoding implementation
+
+        public override bool TryDecode(ReadOnlySpan<byte> data, out string text, out int bytesConsumed)
         {
-            bytesWritten = 0;
-            int justWritten;
-            foreach (var cp in new Utf8String(utf8).CodePoints)
-            {
-                if (!Utf16LittleEndianEncoder.TryEncodeCodePoint(cp, buffer.Slice(bytesWritten), out justWritten))
-                {
-                    bytesWritten = 0;
-                    return false;
-                }
-                bytesWritten += justWritten;
-            }
+            var utf16 = data.Cast<byte, char>();
+            var chars = utf16.ToArray();
+
+            text = new string(chars);
+            bytesConsumed = data.Length;
             return true;
         }
 
-        public override bool TryEncodeFromUtf16(ReadOnlySpan<char> utf16, Span<byte> buffer, out int bytesWritten)
+        public override bool TryDecode(ReadOnlySpan<byte> data, Span<byte> utf8, out int bytesConsumed, out int bytesWritten)
         {
-            var valueBytes = utf16.Cast<char, byte>();
-            if (buffer.Length < valueBytes.Length)
-            {
-                bytesWritten = 0;
-                return false;
-            }
-            valueBytes.CopyTo(buffer);
-            bytesWritten = valueBytes.Length;
-            return true;
+            var utf16 = data.Cast<byte, char>();
+            return Utf8Encoder.TryEncode(utf16, utf8, out bytesConsumed, out bytesWritten);
         }
 
-        public override bool TryEncodeFromUnicode(ReadOnlySpan<UnicodeCodePoint> codePoints, Span<byte> buffer, out int bytesWritten)
+        public override bool TryDecode(ReadOnlySpan<byte> data, Span<char> utf16, out int bytesConsumed, out int charactersWritten)
         {
-            var availableBytes = buffer.Length;
-            var inputLength = codePoints.Length;
-            int bytesWrittenForCodePoint = 0;
-            bytesWritten = 0;
+            // TODO: Other methods validate that the input stream contains valid sequences as they are consumed.
+            //       Currently, this is a copy operation with no validation. What is the right thing here?
```

Comment:
I think no validation is fine for now. If we discover that we need validation, we could add an option to do validation.